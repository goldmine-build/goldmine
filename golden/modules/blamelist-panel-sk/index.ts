import './blamelist-panel-sk';
import './blamelist-panel-sk.scss';
