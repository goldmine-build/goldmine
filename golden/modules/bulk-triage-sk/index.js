import './bulk-triage-sk';
import './bulk-triage-sk.scss';
