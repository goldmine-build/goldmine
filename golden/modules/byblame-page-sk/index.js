import './byblame-page-sk.js';
import './byblame-page-sk.scss';
