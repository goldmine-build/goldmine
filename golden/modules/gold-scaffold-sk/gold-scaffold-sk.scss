@import '~elements-sk/colors';
@import '../colors';

gold-scaffold-sk {
  app-sk {
    --header-background-color: var(--gold);
  }

  * {
    font-family:Roboto,Helvetica,Arial,'Bitstream Vera Sans',sans-serif;
  }

  header {
    h1 {
      font-weight: bold;
      font-size: 1.5em;
      padding: 0 0.25em;
    }

    .spinner-spacer {
      min-width: 40px;
    }

    spinner-sk {
      width: 24px;
      height: 24px;
      margin: 0;
      border: 8px solid var(--light-orange);
      border-left: 8px solid var(--red);
    }
  }

  nav {
    display:flex;
    flex-direction: column;
    a {
      padding: 0.5em 0.75em;
      color: var(--black);
      fill: var(--black);
      text-decoration: none;
    }
    span {
      padding: 0 0.5em;
      vertical-align: middle;
    }
  }

  login-sk {
    margin-left: auto;
    margin-right: 1em;
    display: block;
    color: white;
  }
}
