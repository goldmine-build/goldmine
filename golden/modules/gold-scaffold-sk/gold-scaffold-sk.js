/**
 * @module modules/gold-scaffold-sk
 * @description <h2><code>gold-scaffold-sk</code></h2>
 *
 * Contains the title bar, side bar, and error-toast for all the Gold pages. The rest of
 * every Gold page should be a child of this element.
 *
 * Has a spinner-sk that can be activated when it hears "begin-task" events and keeps
 * spinning until it hears an equal number of "end-task" events.
 *
 * The error-toast element responds to fetch-error events and normal error-sk events.
 *
 * @attr {string} app_title - The title to show in the page banner.
 *
 * @attr {boolean} testing_offline - If we should operate entirely in offline mode.
 */
import { define } from 'elements-sk/define';
import { errorMessage } from 'elements-sk/errorMessage';
import { html } from 'lit-html';
import { ElementSk } from '../../../infra-sk/modules/ElementSk';

import '../../../infra-sk/modules/app-sk';
import '../../../infra-sk/modules/login-sk';

import 'elements-sk/error-toast-sk';
import 'elements-sk/icon/find-in-page-icon-sk';
import 'elements-sk/icon/folder-icon-sk';
import 'elements-sk/icon/help-icon-sk';
import 'elements-sk/icon/home-icon-sk';
import 'elements-sk/icon/label-icon-sk';
import 'elements-sk/icon/laptop-chromebook-icon-sk';
import 'elements-sk/icon/list-icon-sk';
import 'elements-sk/icon/search-icon-sk';
import 'elements-sk/icon/sync-problem-icon-sk';
import 'elements-sk/icon/view-day-icon-sk';
import 'elements-sk/spinner-sk';

const template = (ele) => html`
<app-sk>
  <header>
    <h1>${ele.appTitle}</h1>
    <div class=spinner-spacer>
      <spinner-sk></spinner-sk>
    </div>
    <div class=spacer></div>
    <!-- TODO(kjlubick) last commit -->
    <login-sk ?testing_offline=${ele.testingOffline}></login-sk>
  </header>

  <aside>
    <nav>
      <a href="/" tab-index=0>
        <home-icon-sk></home-icon-sk><span>Home</span>
      </a>
      <a href="/" tab-index=0>
        <view-day-icon-sk></view-day-icon-sk><span>By Blame<span>
      </a>
      <a href="/list" tab-index=0>
        <list-icon-sk></list-icon-sk><span>By Test</span>
      </a>
      <a href="/changelists" tab-index=0>
        <laptop-chromebook-icon-sk></laptop-chromebook-icon-sk><span>By ChangeList</span>
      </a>
      <a href="/search" tab-index=0>
        <search-icon-sk></search-icon-sk><span>Search</span>
      </a>
      <a href="/ignores" tab-index=0>
        <label-icon-sk></label-icon-sk><span>Ignores</span>
      </a>
      <a href="/triagelog" tab-index=0>
        <find-in-page-icon-sk></find-in-page-icon-sk><span>Triage Log</span>
      </a>
      <a href="/failures" tab-index=0>
        <sync-problem-icon-sk></sync-problem-icon-sk><span>Failures</span>
      </a>
      <a href="/help" tab-index=0>
        <help-icon-sk></help-icon-sk><span>Help</span>
      </a>
      <a href="https://github.com/google/skia-buildbot/tree/master/golden" tab-index=0>
        <folder-icon-sk></folder-icon-sk><span>Code</span>
      </a>
    </nav>
  </aside>

  <main></main>

  <footer><error-toast-sk></error-toast-sk></footer>
</app-sk>
`;

/**
 * Moves the elements from one NodeList to another NodeList.
 *
 * @param {NodeList} from - The list we are moving from.
 * @param {NodeList} to - The list we are moving to.
 */
function move(from, to) {
  Array.prototype.slice.call(from).forEach((ele) => to.appendChild(ele));
}

define('gold-scaffold-sk', class extends ElementSk {
  constructor() {
    super(template);
    this._main = null;
    this._busyTaskCount = 0;
    this._spinner = null;
  }

  connectedCallback() {
    super.connectedCallback();
    // Don't call more than once.
    if (this._main) {
      return;
    }
    this.addEventListener('begin-task', this._addBusyTask);
    this.addEventListener('end-task', this._finishedTask);
    this.addEventListener('fetch-error', this._fetchError);

    // We aren't using shadow dom so we need to manually move the children of
    // gold-scaffold-sk to be children of 'main'. We have to do this for the
    // existing elements and for all future mutations.

    // Create a temporary holding spot for elements we're moving.
    const div = document.createElement('div');
    move(this.children, div);

    // Now that we've moved all the old children out of the way we can render
    // the template.
    this._render();

    this._spinner = this.querySelector('header spinner-sk');

    // Move the old children back under main.
    this._main = this.querySelector('main');
    move(div.children, this._main);

    // Move all future children under main also.
    const observer = new MutationObserver((mutList) => {
      mutList.forEach((mut) => {
        move(mut.addedNodes, this._main);
      });
    });
    observer.observe(this, { childList: true });
  }

  disconnectedCallback() {
    super.disconnectedCallback();
    this.removeEventListener('begin-task', this._addBusyTask);
    this.removeEventListener('end-task', this._finishedTask);
    this.removeEventListener('fetch-error', this._fetchError);
  }

  /** @prop appTitle {string} Reflects the app_title attribute for ease of use. */
  get appTitle() { return this.getAttribute('app_title'); }

  set appTitle(val) { this.setAttribute('app_title', val); }

  /** @prop {boolean} busy Indicates if there any on-going tasks (e.g. RPCs).
   *                  This also mirrors the status of the embedded spinner-sk.
   *                  Read-only. */
  get busy() { return !!this._busyTaskCount; }

  /** @prop testingOffline {boolean} Reflects the testing_offline attribute for ease of use.
   */
  get testingOffline() { return this.hasAttribute('testing_offline'); }

  set testingOffline(val) {
    if (val) {
      this.setAttribute('testing_offline', '');
    } else {
      this.removeAttribute('testing_offline');
    }
  }

  /**
   * Indicate there are some number of tasks (e.g. RPCs) the app is waiting on
   * and should be in the "busy" state, if it isn't already.
   *
   */
  _addBusyTask() {
    this._busyTaskCount++;
    if (this._spinner && this._busyTaskCount > 0) {
      this._spinner.active = true;
    }
  }

  /**
   * Removes one task from the busy count. If there are no more tasks to wait
   * for, the app will leave the "busy" state and emit the "busy-end" event.
   *
   */
  _finishedTask() {
    this._busyTaskCount--;
    if (this._busyTaskCount <= 0) {
      this._busyTaskCount = 0;
      if (this._spinner) {
        this._spinner.active = false;
      }
      this.dispatchEvent(new CustomEvent('busy-end', { bubbles: true }));
    }
  }

  /** Handles a fetch error. Expects the detail of error to contain:
   *  error: the error given to fetch.
   *  loading: A string explaining what was being fetched.
   */
  _fetchError(e) {
    const loadingWhat = e.detail.loading;
    e = e.detail.error;
    if (e.name !== 'AbortError') {
      // We can ignore AbortError since they fire anytime an AbortController was canceled.
      // Chrome and Firefox report a DOMException in this case:
      // https://developer.mozilla.org/en-US/docs/Web/API/DOMException
      console.error(e);
      errorMessage(`Unexpected error loading ${loadingWhat}: ${e.message}`,
        5000);
    }
    this._finishedTask();
  }
});
