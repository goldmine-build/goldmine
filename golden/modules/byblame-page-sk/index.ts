import './byblame-page-sk';
import './byblame-page-sk.scss';
