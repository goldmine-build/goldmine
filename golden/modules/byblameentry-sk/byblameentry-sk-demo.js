import './index.js'
import '../gold-scaffold-sk'
import { $$ } from 'common-sk/modules/dom'
import { byBlameEntry, gitLog } from './test_data'

const entry = document.createElement('byblameentry-sk');
entry.byBlameEntry = byBlameEntry;
entry.gitLog = gitLog;
entry.baseRepoUrl = 'https://skia.googlesource.com/skia.git';
entry.corpus = 'gm';
$$('gold-scaffold-sk').appendChild(entry);
