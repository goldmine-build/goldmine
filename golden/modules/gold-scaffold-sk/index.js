import './gold-scaffold-sk';
import './gold-scaffold-sk.scss';
