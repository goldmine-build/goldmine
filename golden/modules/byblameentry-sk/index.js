import './byblameentry-sk.js'
import './byblameentry-sk.scss'
