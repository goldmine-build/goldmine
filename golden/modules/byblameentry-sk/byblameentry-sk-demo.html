<!DOCTYPE html>
<html>
<head>
  <title>byblameentry-sk</title>
  <meta charset="utf-8"/>
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>
<gold-scaffold-sk app_title="Skia Public" testing_offline>
  <!-- byblameentry-sk elements will be inserted here dynamically. -->
</gold-scaffold-sk>
</body>
</html>
